ambry-store/src/test/java/com.github.ambry.store/LogSegmentTest.java

```diff
@@ -350,28 +363,103 @@ public void writeFromTest()
   }
 
   /**
+   * Tests for special constructor cases.
+   * @throws IOException
+   */
+  @Test
+  public void constructorTest()
+      throws IOException {
+    LogSegment segment = getSegment("log_current", STANDARD_SEGMENT_SIZE, false);
+    assertEquals("Start offset should be 0 when no headers are written", 0, segment.getStartOffset());
+  }
+
+  /**
    * Tests for bad construction cases of {@link LogSegment}.
    * @throws IOException
    */
   @Test
   public void badConstructionTest()
       throws IOException {
     // try to construct with a file that does not exist.
+    String name = "log_non_existent";
+    File file = new File(tempDir, name);
+    try {
+      new LogSegment(name, file, STANDARD_SEGMENT_SIZE, metrics, true);
+      fail("Construction should have failed because the backing file does not exist");
+    } catch (IllegalArgumentException e) {
+      // expected. Nothing to do.
+    }
+
     try {
-      new LogSegment("log_non_existent", new File(tempDir, "log_non_existent"), STANDARD_SEGMENT_SIZE, metrics);
+      new LogSegment(name, file, metrics);
       fail("Construction should have failed because the backing file does not exist");
     } catch (IllegalArgumentException e) {
       // expected. Nothing to do.
     }
 
     // try to construct with a file that is a directory
+    name = tempDir.getName();
+    file = new File(tempDir.getParent(), name);
     try {
-      new LogSegment(tempDir.getName(), new File(tempDir.getParent(), tempDir.getName()), STANDARD_SEGMENT_SIZE,
-          metrics);
+      new LogSegment(name, file, STANDARD_SEGMENT_SIZE, metrics, true);
       fail("Construction should have failed because the backing file does not exist");
     } catch (IllegalArgumentException e) {
       // expected. Nothing to do.
     }
+
+    name = tempDir.getName();
+    file = new File(tempDir.getParent(), name);
+    try {
+      new LogSegment(name, file, metrics);
+      fail("Construction should have failed because the backing file does not exist");
+    } catch (IllegalArgumentException e) {
+      // expected. Nothing to do.
+    }
+
+    // unknown version
+    LogSegment segment = getSegment("log_current", STANDARD_SEGMENT_SIZE, true);
+    file = segment.getView().getFirst();
+    byte[] header = getHeader(segment);
+    byte savedByte = header[0];
+    // mess with version
+    header[0] = (byte) (header[0] + 10);
+    writeHeader(segment, header);
+    try {
+      new LogSegment(name, file, metrics);
+      fail("Construction should have failed because version is unknown");
+    } catch (IllegalArgumentException e) {
+      // expected. Nothing to do.
+    }
+
+    // bad CRC
+    // fix version but mess with data after version
+    header[0] = savedByte;
+    header[2] = header[2] == (byte) 1 ? (byte) 0 : (byte) 1;
+    writeHeader(segment, header);
+    try {
+      new LogSegment(name, file, metrics);
+      fail("Construction should have failed because crc check should have failed");
```

Comment:
Not sure what the point of corrupting the CRC is? The test here is to check that we do a CRC match - this means that the data has to be corrupted, not the CRC.